From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Thu, 31 Oct 2024 23:22:33 +0800
Subject: [PATCH] Setup publishing


diff --git a/build.gradle b/build.gradle
index ca132d92..7d867b56 100644
--- a/build.gradle
+++ b/build.gradle
@@ -284,7 +284,7 @@ apply plugin: 'me.modmuss50.mod-publish-plugin'
 
 publishMods {
     changelog = ghaChangelog
-    modLoaders.add("fabric")
+    modLoaders.add("neoforge")
 
     def envDisplayName = System.getenv("OVERRIDE_DISPLAY_NAME")
     displayName = envDisplayName != null ? envDisplayName : (project.version + " devbuild for " + project.minecraft_version)
@@ -321,7 +321,7 @@ publishMods {
         modrinth("modrinthMITPackage") {
             from mrOptions
             file = remapJar.archiveFile
-            projectId = "VSNURh3q"
+            projectId = "COlSi5iR"
         }
     }
 }
