From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Wed, 24 Sep 2025 21:32:48 +0800
Subject: [PATCH] new: delayFullChunkEvents


diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/ChunkState.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/ChunkState.java
index 50e44542..2e905ad0 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/ChunkState.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/ChunkState.java
@@ -4,5 +4,5 @@ import net.minecraft.world.chunk.Chunk;
 import net.minecraft.world.chunk.ChunkStatus;
 import net.minecraft.world.chunk.ProtoChunk;
 
-public record ChunkState(Chunk chunk, ProtoChunk protoChunk, ChunkStatus reachedStatus) {
+public record ChunkState(Chunk chunk, ProtoChunk protoChunk, ChunkStatus reachedStatus, boolean wasFullChunk) {
 }
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
index 2a92783d..37bf569a 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
@@ -101,6 +101,16 @@ public class Config {
                     """)
             .getBoolean(false, false);
 
+    public static final boolean delayFullChunkEvents = new ConfigSystem.ConfigAccessor()
+            .key("chunkSystem.delayFullChunkEvents")
+            .comment("""
+                    Whether to delay full chunk events to after full chunk future have been completed
+                    This restores the buggy behavior in NeoForge that the ecosystem have been relying on
+                    
+                    Do not change this unless you know what you are doing
+                    """)
+            .getBoolean(true, false);
+
     public static void init() {
         // intentionally empty
     }
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/TheChunkSystem.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/TheChunkSystem.java
index fd192de0..4fdbedbc 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/TheChunkSystem.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/TheChunkSystem.java
@@ -90,7 +90,7 @@ public class TheChunkSystem extends StatusAdvancingScheduler<ChunkPos, ChunkStat
     protected void onItemCreation(ItemHolder<ChunkPos, ChunkState, ChunkLoadingContext, NewChunkHolderVanillaInterface> holder) {
         super.onItemCreation(holder);
         holder.getUserData().set(new NewChunkHolderVanillaInterface(this, holder, ((IThreadedAnvilChunkStorage) this.tacs).getWorld(), ((IThreadedAnvilChunkStorage) this.tacs).getLightingProvider(), this.tacs));
-        holder.getItem().set(new ChunkState(null, null, null));
+        holder.getItem().set(new ChunkState(null, null, null, false));
     }
 
     @Override
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
index fe5fdc5d..c306cce2 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
@@ -80,7 +80,7 @@ public class ReadFromDisk extends NewChunkStatus {
                     }
                 })
                 .doOnSuccess(chunk -> {
-                    context.holder().getItem().set(new ChunkState(chunk, chunk, ChunkStatus.EMPTY));
+                    context.holder().getItem().set(new ChunkState(chunk, chunk, ChunkStatus.EMPTY, chunk instanceof WrapperProtoChunk));
                     if (!Config.lowMemoryMode) {
                         context.holder().getUserData().get().triggerDeferredLoad(NewChunkStatus.DISK);
                     }
@@ -217,7 +217,7 @@ public class ReadFromDisk extends NewChunkStatus {
 
                         ((IPOIUnloading) ((IThreadedAnvilChunkStorage) context.tacs()).getPointOfInterestStorage()).c2me$unloadPoi(context.holder().getKey());
 
-                        context.holder().getItem().set(new ChunkState(null, null, null));
+                        context.holder().getItem().set(new ChunkState(null, null, null, false));
 
                         return asyncSaveFuture;
                     }
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerAccessible.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerAccessible.java
index 3e276ba0..06b0d269 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerAccessible.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerAccessible.java
@@ -6,6 +6,7 @@ import com.ishland.c2me.base.mixin.access.IThreadedAnvilChunkStorage;
 import com.ishland.c2me.base.mixin.access.IWorldChunk;
 import com.ishland.c2me.rewrites.chunksystem.common.ChunkLoadingContext;
 import com.ishland.c2me.rewrites.chunksystem.common.ChunkState;
+import com.ishland.c2me.rewrites.chunksystem.common.Config;
 import com.ishland.c2me.rewrites.chunksystem.common.NewChunkStatus;
 import com.ishland.c2me.rewrites.chunksystem.common.compat.lithium.LithiumChunkStatusTrackerInvoker;
 import com.ishland.c2me.rewrites.chunksystem.common.ducks.WorldChunkExtension;
@@ -55,19 +56,10 @@ public class ServerAccessible extends NewChunkStatus {
                         worldChunk.setLevelTypeProvider(context.holder().getUserData().get()::getLevelType);
                         worldChunk.setUnsavedListener(((IThreadedAnvilChunkStorage) context.tacs()).getGenerationContext().unsavedListener());
                         ((WorldChunkExtension) worldChunk).c2me$setBlockTicking(false); // not necessary, but just in case
-                        context.holder().getItem().set(new ChunkState(worldChunk, new WrapperProtoChunk(worldChunk, false), ChunkStatus.FULL));
-                        if (!((IWorldChunk) worldChunk).isLoadedToWorld()) {
-                            worldChunk.loadEntities();
-                            worldChunk.setLoadedToWorld(true);
-                            worldChunk.updateAllBlockEntities();
-                            worldChunk.addChunkTickSchedulers(serverWorld);
-                            LifecycleEventInvoker.invokeChunkLoaded(serverWorld, worldChunk, !(protoChunk instanceof WrapperProtoChunk));
-                }
-
-                        LifecycleEventInvoker.invokeChunkLevelTypeChange(serverWorld, worldChunk, ChunkLevelType.INACCESSIBLE, ChunkLevelType.FULL);
-
-                        ((IThreadedAnvilChunkStorage) context.tacs()).getCurrentChunkHolders().put(context.holder().getKey().toLong(), context.holder().getUserData().get());
-                        ((IThreadedAnvilChunkStorage) context.tacs()).setChunkHolderListDirty(true);
+                        context.holder().getItem().set(new ChunkState(worldChunk, new WrapperProtoChunk(worldChunk, false), ChunkStatus.FULL, protoChunk instanceof WrapperProtoChunk));
+                        if (!Config.delayFullChunkEvents) {
+                            invokeFullChunkEvents(context, worldChunk, serverWorld, protoChunk instanceof WrapperProtoChunk);
+                        }
                     }
                 })
                 .subscribeOn(Schedulers.from(((IThreadedAnvilChunkStorage) context.tacs()).getMainThreadExecutor()));
@@ -75,7 +67,31 @@ public class ServerAccessible extends NewChunkStatus {
 
     @Override
     public Completable postUpgradeToThis(ChunkLoadingContext context) {
-        return Completable.complete();
+        if (!Config.delayFullChunkEvents) return Completable.complete();
+
+        return Completable.fromRunnable(() -> {
+            Assertions.assertTrue(((IThreadedAnvilChunkStorage) context.tacs()).getMainThreadExecutor().isOnThread());
+            ChunkState chunkState = context.holder().getItem().get();
+            WorldChunk worldChunk = (WorldChunk) chunkState.chunk();
+            ServerWorld serverWorld = ((IThreadedAnvilChunkStorage) context.tacs()).getWorld();
+            invokeFullChunkEvents(context, worldChunk, serverWorld, chunkState.wasFullChunk());
+        });
+    }
+
+    private static void invokeFullChunkEvents(ChunkLoadingContext context, WorldChunk worldChunk, ServerWorld serverWorld, boolean wasFullChunk) {
+        if (!((IWorldChunk) worldChunk).isLoadedToWorld()) {
+            worldChunk.loadEntities();
+            worldChunk.setLoadedToWorld(true);
+            worldChunk.updateAllBlockEntities();
+            worldChunk.addChunkTickSchedulers(serverWorld);
+            LifecycleEventInvoker.invokeChunkLoaded(serverWorld, worldChunk, !wasFullChunk);
+        }
+
+
+        LifecycleEventInvoker.invokeChunkLevelTypeChange(serverWorld, worldChunk, ChunkLevelType.INACCESSIBLE, ChunkLevelType.FULL);
+
+        ((IThreadedAnvilChunkStorage) context.tacs()).getCurrentChunkHolders().put(context.holder().getKey().toLong(), context.holder().getUserData().get());
+        ((IThreadedAnvilChunkStorage) context.tacs()).setChunkHolderListDirty(true);
     }
 
     private static WorldChunk toFullChunk(ProtoChunk protoChunk, ServerWorld serverWorld) {
@@ -121,7 +137,7 @@ public class ServerAccessible extends NewChunkStatus {
                         worldChunk.setLevelTypeProvider(null);
                         worldChunk.setUnsavedListener(pos -> {
                         });
-                        context.holder().getItem().set(new ChunkState(state.protoChunk(), state.protoChunk(), ChunkStatus.FULL));
+                        context.holder().getItem().set(new ChunkState(state.protoChunk(), state.protoChunk(), ChunkStatus.FULL, true));
                     }
                 })
                 .subscribeOn(Schedulers.from(((IThreadedAnvilChunkStorage) context.tacs()).getMainThreadExecutor()));
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/VanillaWorldGenerationDelegate.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/VanillaWorldGenerationDelegate.java
index 8f9e99cc..9c17a371 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/VanillaWorldGenerationDelegate.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/VanillaWorldGenerationDelegate.java
@@ -24,6 +24,7 @@ import net.minecraft.world.chunk.ChunkGenerationSteps;
 import net.minecraft.world.chunk.ChunkStatus;
 import net.minecraft.world.chunk.GenerationDependencies;
 import net.minecraft.world.chunk.ProtoChunk;
+import net.minecraft.world.chunk.WrapperProtoChunk;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -123,7 +124,7 @@ public class VanillaWorldGenerationDelegate extends NewChunkStatus {
                                 .run(((IThreadedAnvilChunkStorage) context.tacs()).getGenerationContext(), context.chunks(), chunk)
                                 .whenComplete((chunk1, throwable) -> {
                                     if (chunk1 != null) {
-                                        context.holder().getItem().set(new ChunkState(chunk1, (ProtoChunk) chunk1, this.status));
+                                        context.holder().getItem().set(new ChunkState(chunk1, (ProtoChunk) chunk1, this.status, chunk1 instanceof WrapperProtoChunk));
                                     }
                                 })
                 ));
@@ -138,7 +139,7 @@ public class VanillaWorldGenerationDelegate extends NewChunkStatus {
                             return step.run(chunkGenerationContext, context.chunks(), chunk)
                                     .whenComplete((chunk1, throwable) -> {
                                         if (chunk1 != null) {
-                                            context.holder().getItem().set(new ChunkState(chunk1, (ProtoChunk) chunk1, this.status));
+                                            context.holder().getItem().set(new ChunkState(chunk1, (ProtoChunk) chunk1, this.status, chunk1 instanceof WrapperProtoChunk));
                                         }
                                     }).thenAccept(__ -> {
                                     });
