From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Sat, 8 Nov 2025 23:36:25 +0800
Subject: [PATCH] fix: stop-gap solution for create virtual render world


diff --git a/c2me-fixes-worldgen-threading-issues/src/main/java/com/ishland/c2me/fixes/worldgen/threading_issues/mixin/threading_detections/random_instances/MixinWorld.java b/c2me-fixes-worldgen-threading-issues/src/main/java/com/ishland/c2me/fixes/worldgen/threading_issues/mixin/threading_detections/random_instances/MixinWorld.java
index cc549c1b..a695f6a9 100644
--- a/c2me-fixes-worldgen-threading-issues/src/main/java/com/ishland/c2me/fixes/worldgen/threading_issues/mixin/threading_detections/random_instances/MixinWorld.java
+++ b/c2me-fixes-worldgen-threading-issues/src/main/java/com/ishland/c2me/fixes/worldgen/threading_issues/mixin/threading_detections/random_instances/MixinWorld.java
@@ -42,6 +42,13 @@ public class MixinWorld {
     @Redirect(method = "<init>", at = @At(value = "INVOKE", target = "Lnet/minecraft/util/math/random/Random;create()Lnet/minecraft/util/math/random/Random;"))
     private Random redirectWorldRandomInit() {
 //        new CheckedThreadLocalRandom(RandomSeed.getSeed(), () -> new Thread()).nextInt();
+        Class<?> clazz = this.getClass();
+        while (clazz != null) {
+            if (clazz.getName().equals("com.simibubi.create.foundation.virtualWorld.VirtualRenderWorld")) { // stop-gap solution
+                return Random.create();
+            }
+            clazz = clazz.getSuperclass();
+        }
         return new CheckedThreadLocalRandom(RandomSeed.getSeed(), () -> this.thread);
     }
 
