From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Thu, 17 Jul 2025 23:54:31 +0800
Subject: [PATCH] fix: NeoForge dumpster fire


diff --git a/src/main/java/com/ishland/c2me/PreLaunchHandler.java b/src/main/java/com/ishland/c2me/PreLaunchHandler.java
index b57175bd..f223328a 100644
--- a/src/main/java/com/ishland/c2me/PreLaunchHandler.java
+++ b/src/main/java/com/ishland/c2me/PreLaunchHandler.java
@@ -35,7 +35,11 @@ public class PreLaunchHandler {
                     }
                     for (String s : unhandled) {
                         auditLogger.info("Loading class {}", s);
-                        MixinService.getService().getClassProvider().findClass(s, false);
+                        try {
+                            MixinService.getService().getClassProvider().findClass(s, false);
+                        } catch (ClassNotFoundException e) {
+                            auditLogger.warn("Failed to find class {}", s, e);
+                        }
                     }
                     for (Object config : configs) {
                         for (String unhandledTarget : (Set<String>) getUnhandledTargetsMethod.invoke(config)) {
