From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Sat, 10 Jan 2026 17:06:09 +0800
Subject: [PATCH] change: mark useEndBiomeCache imcompatible with
 simplyimprovedterrain


diff --git a/c2me-opts-worldgen-vanilla/src/main/java/com/ishland/c2me/opts/worldgen/vanilla/common/Config.java b/c2me-opts-worldgen-vanilla/src/main/java/com/ishland/c2me/opts/worldgen/vanilla/common/Config.java
index 346673cf..4096c0f0 100644
--- a/c2me-opts-worldgen-vanilla/src/main/java/com/ishland/c2me/opts/worldgen/vanilla/common/Config.java
+++ b/c2me-opts-worldgen-vanilla/src/main/java/com/ishland/c2me/opts/worldgen/vanilla/common/Config.java
@@ -19,6 +19,7 @@ public class Config {
                     (may cause incompatibility with other mods)
                     """)
             .incompatibleMod("biolith", "*")
+            .incompatibleMod("simplyimprovedterrain", "*")
             .getBoolean(true, false);
 
     public static final boolean optimizeStructureWeightSampler = new ConfigSystem.ConfigAccessor()
