From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Sun, 1 Mar 2026 12:08:18 +0800
Subject: [PATCH] change: drop object_pooling_caching.MixinConfiguredFeature

Cherry-picked from 78b2733e06bb22770fdf03dffe94b6e8f7e3e889

diff --git a/c2me-opts-allocs/src/main/java/com/ishland/c2me/opts/allocs/mixin/object_pooling_caching/MixinConfiguredFeature.java b/c2me-opts-allocs/src/main/java/com/ishland/c2me/opts/allocs/mixin/object_pooling_caching/MixinConfiguredFeature.java
deleted file mode 100644
index 8730c4aa..00000000
--- a/c2me-opts-allocs/src/main/java/com/ishland/c2me/opts/allocs/mixin/object_pooling_caching/MixinConfiguredFeature.java
+++ /dev/null
@@ -1,43 +0,0 @@
-package com.ishland.c2me.opts.allocs.mixin.object_pooling_caching;
-
-import com.ishland.c2me.opts.allocs.common.PooledFeatureContext;
-import com.ishland.flowsched.structs.SimpleObjectPool;
-import net.minecraft.util.math.BlockPos;
-import net.minecraft.util.math.random.Random;
-import net.minecraft.world.StructureWorldAccess;
-import net.minecraft.world.gen.chunk.ChunkGenerator;
-import net.minecraft.world.gen.feature.ConfiguredFeature;
-import net.minecraft.world.gen.feature.Feature;
-import net.minecraft.world.gen.feature.FeatureConfig;
-import org.spongepowered.asm.mixin.Final;
-import org.spongepowered.asm.mixin.Mixin;
-import org.spongepowered.asm.mixin.Overwrite;
-import org.spongepowered.asm.mixin.Shadow;
-
-import java.util.Optional;
-
-@Mixin(ConfiguredFeature.class)
-public class MixinConfiguredFeature<FC extends FeatureConfig, F extends Feature<FC>> {
-
-    @Shadow @Final public F feature;
-
-    @Shadow @Final public FC config;
-
-    /**
-     * @author ishland
-     * @reason pool FeatureContext
-     */
-    @Overwrite
-    public boolean generate(StructureWorldAccess world, ChunkGenerator chunkGenerator, Random random, BlockPos origin) {
-        if (!world.isValidForSetBlock(origin)) return false;
-        final SimpleObjectPool<PooledFeatureContext<?>> pool = PooledFeatureContext.POOL.get();
-        final PooledFeatureContext<FC> context = (PooledFeatureContext<FC>) pool.alloc();
-        try {
-            context.reInit(Optional.empty(), world, chunkGenerator, random, origin, this.config);
-            return this.feature.generate(context);
-        } finally {
-            pool.release(context);
-        }
-    }
-
-}
diff --git a/c2me-opts-allocs/src/main/resources/c2me-opts-allocs.mixins.json b/c2me-opts-allocs/src/main/resources/c2me-opts-allocs.mixins.json
index 1e41f49e..398f80e3 100644
--- a/c2me-opts-allocs/src/main/resources/c2me-opts-allocs.mixins.json
+++ b/c2me-opts-allocs/src/main/resources/c2me-opts-allocs.mixins.json
@@ -8,7 +8,6 @@
     "MixinUtil",
     "asm.ASMTargets",
     "noise.MixinChainedBlockSource",
-    "object_pooling_caching.MixinConfiguredFeature",
     "object_pooling_caching.MixinOreFeature",
     "surfacebuilder.MixinMaterialRuleContext",
     "surfacebuilder.MixinMaterialRulesSequenceBlockStateRule",
