From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Wed, 28 Jan 2026 11:36:46 +0800
Subject: [PATCH] fix: prevent item drops on worldgen threads


diff --git a/c2me-fixes-worldgen-threading-issues/src/main/java/com/ishland/c2me/fixes/worldgen/threading_issues/mixin/threading/MixinChunkRegion.java b/c2me-fixes-worldgen-threading-issues/src/main/java/com/ishland/c2me/fixes/worldgen/threading_issues/mixin/threading/MixinChunkRegion.java
new file mode 100644
index 00000000..b4b6e872
--- /dev/null
+++ b/c2me-fixes-worldgen-threading-issues/src/main/java/com/ishland/c2me/fixes/worldgen/threading_issues/mixin/threading/MixinChunkRegion.java
@@ -0,0 +1,39 @@
+package com.ishland.c2me.fixes.worldgen.threading_issues.mixin.threading;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.ChunkRegion;
+import net.minecraft.world.chunk.ChunkGenerationStep;
+import org.slf4j.Logger;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
+
+import java.util.function.Supplier;
+
+@Mixin(ChunkRegion.class)
+public class MixinChunkRegion {
+
+    @Shadow
+    @Final
+    private static Logger LOGGER;
+
+    @Shadow
+    @Final
+    private ChunkGenerationStep generationStep;
+
+    @Shadow
+    private Supplier<String> currentlyGeneratingStructureName;
+
+    @ModifyVariable(method = "breakBlock", at = @At("HEAD"), argsOnly = true)
+    private boolean preventDropItem(final boolean drop, final BlockPos pos, final boolean drop1, final Entity breakingEntity, final int maxUpdateDepth) {
+        if (drop) {
+            LOGGER.error("Detected breakBlock item drop on pos {}, status: {}, currently generating: {}",
+                    pos, this.generationStep.targetStatus(), this.currentlyGeneratingStructureName == null ? "unknown": this.currentlyGeneratingStructureName.get());
+        }
+        return false;
+    }
+
+}
diff --git a/c2me-fixes-worldgen-threading-issues/src/main/resources/c2me-fixes-worldgen-threading-issues.mixins.json b/c2me-fixes-worldgen-threading-issues/src/main/resources/c2me-fixes-worldgen-threading-issues.mixins.json
index 5979b61c..df5c18e8 100644
--- a/c2me-fixes-worldgen-threading-issues/src/main/resources/c2me-fixes-worldgen-threading-issues.mixins.json
+++ b/c2me-fixes-worldgen-threading-issues/src/main/resources/c2me-fixes-worldgen-threading-issues.mixins.json
@@ -6,6 +6,7 @@
   "mixins": [
     "deadlocks.MixinDataFixerType",
     "threading.MixinChunk",
+    "threading.MixinChunkRegion",
     "threading.MixinDesertTempleGenerator",
     "threading.MixinJungleTempleGenerator",
     "threading.MixinMineshaftGeneratorMineshaftCorridor",
