From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Tue, 1 Apr 2025 13:23:16 +0800
Subject: [PATCH] change: lower block entity warning to debug


diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationManager.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationManager.java
index b21da3b9..c82c5b40 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationManager.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/async_chunkio/AsyncSerializationManager.java
@@ -53,7 +53,8 @@ public class AsyncSerializationManager {
             for (BlockPos blockPos : chunk.getBlockEntityPositions()) {
                 final NbtCompound nbt = chunk.getPackedBlockEntityNbt(blockPos, world.getRegistryManager());
                 if (nbt == null) {
-                    LOGGER.warn("Block entity at {} for block {} in chunk {} is missing", blockPos, chunk.getBlockState(blockPos), chunk.getPos());
+                    // the game actually allows this to exist
+                    LOGGER.debug("Block entity at {} for block {} in chunk {} is missing", blockPos, chunk.getBlockState(blockPos), chunk.getPos());
                 }
                 if (blockEntities.containsKey(blockPos)) {
                     LOGGER.warn("Duplicate block entity at {} in chunk {}", blockPos, chunk.getPos());
