From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Sat, 19 Apr 2025 19:28:28 +0800
Subject: [PATCH] perf: filterFluidPostProcessing


diff --git a/c2me-base/src/main/java/com/ishland/c2me/base/mixin/access/IFlowableFluid.java b/c2me-base/src/main/java/com/ishland/c2me/base/mixin/access/IFlowableFluid.java
new file mode 100644
index 00000000..065866a3
--- /dev/null
+++ b/c2me-base/src/main/java/com/ishland/c2me/base/mixin/access/IFlowableFluid.java
@@ -0,0 +1,26 @@
+package com.ishland.c2me.base.mixin.access;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.fluid.FlowableFluid;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.BlockView;
+import net.minecraft.world.WorldView;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+@Mixin(FlowableFluid.class)
+public interface IFlowableFluid {
+
+    @Invoker
+    int invokeGetLevelDecreasePerBlock(WorldView world);
+
+    @Invoker
+    boolean invokeReceivesFlow(Direction face, BlockView world, BlockPos pos, BlockState state, BlockPos fromPos, BlockState fromState);
+
+    @Invoker
+    boolean invokeIsMatchingAndStill(FluidState state);
+
+}
diff --git a/c2me-base/src/main/resources/c2me-base.mixins.json b/c2me-base/src/main/resources/c2me-base.mixins.json
index 2f5d0395..3362f909 100644
--- a/c2me-base/src/main/resources/c2me-base.mixins.json
+++ b/c2me-base/src/main/resources/c2me-base.mixins.json
@@ -22,6 +22,7 @@
     "access.IChunkTickScheduler",
     "access.IDensityFunctionsCaveScaler",
     "access.IDensityFunctionTypesWeirdScaledSamplerRarityValueMapper",
+    "access.IFlowableFluid",
     "access.IInterpolatedNoiseSampler",
     "access.IMultiNoiseBiomeSource",
     "access.INbtCompound",
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
index c3a7c5c7..48a8f9b8 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
@@ -69,6 +69,27 @@ public class Config {
                     """)
             .getBoolean(true, false);
 
+    public static final boolean filterFluidPostProcessing = new ConfigSystem.ConfigAccessor()
+            .key("chunkSystem.filterFluidPostProcessing")
+            .comment("""
+                    Whether to filter fluid post-processing on worldgen threads
+                    
+                    The worldgen processes creates a lot of unnecessary fluid post-processing tasks,
+                    which can overload the server thread and cause stutters.
+                    This applies a rough filter to filter out fluids that are definitely not going to flow
+                    """)
+            .getBoolean(true, false);
+
+    public static final boolean lowMemoryMode = new ConfigSystem.ConfigAccessor()
+            .key("chunkSystem.lowMemoryMode")
+            .comment("""
+                    Whether to enable low memory mode
+                    
+                    This will attempt to aggressively unload unneeded chunks, saving memory at the cost of additional
+                    overhead when generating new chunks.
+                    """)
+            .getBoolean(false, false);
+
     public static void init() {
         // intentionally empty
     }
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/quirks/FlowableFluidUtils.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/quirks/FlowableFluidUtils.java
new file mode 100644
index 00000000..1960ffdd
--- /dev/null
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/quirks/FlowableFluidUtils.java
@@ -0,0 +1,161 @@
+package com.ishland.c2me.rewrites.chunksystem.common.quirks;
+
+import com.ishland.c2me.base.mixin.access.IFlowableFluid;
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.block.DoorBlock;
+import net.minecraft.block.FluidFillable;
+import net.minecraft.fluid.FlowableFluid;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.registry.tag.BlockTags;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.BlockView;
+import net.minecraft.world.WorldView;
+
+public class FlowableFluidUtils {
+
+    public static final ThreadLocal<Boolean> shouldntFlow = new ThreadLocal<>();
+    public static final ThreadLocal<Boolean> hadFlown = new ThreadLocal<>();
+
+    public static boolean needsPostProcessing(WorldView world, BlockPos pos, BlockState blockState, FluidState fluidState) {
+        if (!fluidState.isStill()) {
+            return true;
+        }
+        return canFlowNormally(world, pos, blockState, fluidState);
+    }
+
+    private static boolean canFlowNormally(WorldView world, BlockPos pos, BlockState blockState, FluidState fluidState) {
+        if (fluidState.isEmpty()) return false;
+
+        BlockPos belowPos = pos.down();
+        BlockState belowBlockState = world.getBlockState(belowPos);
+        FluidState belowFluidState = world.getFluidState(belowPos);
+        // very rough filtering
+        if (canFlowThrough1_21_5((FlowableFluid) fluidState.getFluid(), world, pos, blockState, Direction.DOWN, belowPos, belowBlockState, belowFluidState)) {
+            FluidState fluidState3 = getUpdatedState(((FlowableFluid) fluidState.getFluid()), world, belowPos, belowBlockState);
+            if (fluidState3 == null) {
+                return true; // shortcut
+            }
+            Fluid fluid = fluidState3.getFluid();
+            if (belowFluidState.canBeReplacedWith(world, belowPos, fluid, Direction.DOWN) && canFillWithFluid1_21_5(world, belowPos, belowBlockState, fluid)) {
+                return true;
+            }
+        }
+        if (canSpreadToSidesNormally(world, pos, blockState, fluidState)) { // fluid always still when reached here
+            return true;
+        }
+
+        return false;
+    }
+
+    private static boolean canSpreadToSidesNormally(WorldView world, BlockPos pos, BlockState blockState, FluidState fluidState) {
+        int nextFluidLevel = fluidState.getLevel() - ((IFlowableFluid) fluidState.getFluid()).invokeGetLevelDecreasePerBlock(world);
+        if (fluidState.get(FlowableFluid.FALLING)) {
+            nextFluidLevel = 7;
+        }
+        if (nextFluidLevel > 0) {
+            // getSpread
+//            int i = 1000;
+//            Map<Direction, FluidState> map = Maps.newEnumMap(Direction.class);
+//            SpreadCache spreadCache = null;
+
+            for (Direction direction : Direction.Type.HORIZONTAL) {
+                BlockPos offsetPos = pos.offset(direction);
+                BlockState offsetBlockState = world.getBlockState(offsetPos);
+                FluidState offsetFluidState = offsetBlockState.getFluidState();
+                if (canFlowThrough1_21_5((FlowableFluid) fluidState.getFluid(), world, pos, blockState, direction, offsetPos, offsetBlockState, offsetFluidState)) {
+//                    FluidState fluidState2 = getUpdatedState((FlowableFluid) fluidState.getFluid(), world, offsetPos, offsetBlockState);
+//                    if (fluidState2 == null) {
+//                        return true; // shortcut
+//                    }
+//                    if (canFillWithFluid1_21_5(world, offsetPos, offsetBlockState, fluidState2.getFluid())) {
+//                        return true; // shortcut
+//                    }
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean canFlowThrough1_21_5(FlowableFluid receiver, BlockView world, BlockPos pos, BlockState state, Direction face, BlockPos fromPos, BlockState fromState, FluidState fluidState) {
+        return !((IFlowableFluid) receiver).invokeIsMatchingAndStill(fluidState) &&
+                canFillShort(fromState) &&
+                ((IFlowableFluid) receiver).invokeReceivesFlow(face, world, pos, state, fromPos, fromState);
+    }
+
+    private static boolean canFill(BlockView world, BlockPos pos, BlockState state, Fluid fluid) {
+        return canFillShort(state) && canFillWithFluid1_21_5(world, pos, state, fluid);
+    }
+
+    private static boolean canFillWithFluid1_21_5(BlockView world, BlockPos pos, BlockState state, Fluid fluid) {
+        return state.getBlock() instanceof FluidFillable fluidFillable ? fluidFillable.canFillWithFluid(null, world, pos, state, fluid) : true;
+    }
+
+    private static boolean canFlowDownTo(FlowableFluid receiver, BlockView world, BlockPos pos, BlockState state, BlockPos fromPos, BlockState fromState) {
+        if (!((IFlowableFluid) receiver).invokeReceivesFlow(Direction.DOWN, world, pos, state, fromPos, fromState)) {
+            return false;
+        } else {
+            return fromState.getFluidState().getFluid().matchesType(receiver) ? true : canFill(world, fromPos, fromState, receiver.getFlowing());
+        }
+    }
+
+    private static boolean canFillShort(BlockState state) {
+        Block block = state.getBlock();
+        if (block instanceof FluidFillable) {
+            return true; // shortcut
+        } else if (!(block instanceof DoorBlock)
+                && !state.isIn(BlockTags.SIGNS)
+                && !state.isOf(Blocks.LADDER)
+                && !state.isOf(Blocks.SUGAR_CANE)
+                && !state.isOf(Blocks.BUBBLE_COLUMN)) {
+            return !state.isOf(Blocks.NETHER_PORTAL) &&
+                    !state.isOf(Blocks.END_PORTAL) &&
+                    !state.isOf(Blocks.END_GATEWAY) &&
+                    !state.isOf(Blocks.STRUCTURE_VOID)
+                    ? !state.blocksMovement()
+                    : false;
+        } else {
+            return false;
+        }
+    }
+
+    private static FluidState getUpdatedState(FlowableFluid receiver, WorldView world, BlockPos pos, BlockState state) {
+        int i = 0;
+        int j = 0;
+        BlockPos.Mutable mutable = new BlockPos.Mutable();
+
+        for (Direction direction : Direction.Type.HORIZONTAL) {
+            BlockPos blockPos = mutable.set(pos, direction);
+            BlockState blockState = world.getBlockState(blockPos);
+            FluidState fluidState = blockState.getFluidState();
+            if (fluidState.getFluid().matchesType(receiver) && ((IFlowableFluid) receiver).invokeReceivesFlow(direction, world, pos, state, blockPos, blockState)) {
+                if (fluidState.isStill()) {
+                    j++;
+                }
+
+                i = Math.max(i, fluidState.getLevel());
+            }
+        }
+
+        if (j >= 2) {
+            return null; // to not filter this
+        }
+
+        BlockPos blockPos2 = mutable.set(pos, Direction.UP);
+        BlockState blockState3 = world.getBlockState(blockPos2);
+        FluidState fluidState3 = blockState3.getFluidState();
+        if (!fluidState3.isEmpty() && fluidState3.getFluid().matchesType(receiver) && ((IFlowableFluid) receiver).invokeReceivesFlow(Direction.UP, world, pos, state, blockPos2, blockState3)) {
+            return receiver.getFlowing(8, true);
+        } else {
+            int k = i - ((IFlowableFluid) receiver).invokeGetLevelDecreasePerBlock(world);
+            return k <= 0 ? Fluids.EMPTY.getDefaultState() : receiver.getFlowing(k, false);
+        }
+    }
+
+}
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerBlockTicking.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerBlockTicking.java
index db208837..02cbb716 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerBlockTicking.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ServerBlockTicking.java
@@ -1,20 +1,33 @@
 package com.ishland.c2me.rewrites.chunksystem.common.statuses;
 
+import com.google.common.base.Suppliers;
 import com.ishland.c2me.base.common.threadstate.ThreadInstrumentation;
 import com.ishland.c2me.base.mixin.access.IThreadedAnvilChunkStorage;
 import com.ishland.c2me.rewrites.chunksystem.common.ChunkLoadingContext;
 import com.ishland.c2me.rewrites.chunksystem.common.ChunkState;
+import com.ishland.c2me.rewrites.chunksystem.common.Config;
 import com.ishland.c2me.rewrites.chunksystem.common.NewChunkStatus;
+import com.ishland.c2me.rewrites.chunksystem.common.quirks.FlowableFluidUtils;
 import com.ishland.c2me.rewrites.chunksystem.common.threadstate.ChunkTaskWork;
 import com.ishland.flowsched.scheduler.Cancellable;
 import com.ishland.flowsched.scheduler.ItemHolder;
 import com.ishland.flowsched.scheduler.KeyStatusPair;
+import it.unimi.dsi.fastutil.shorts.ShortList;
+import it.unimi.dsi.fastutil.shorts.ShortListIterator;
+import net.minecraft.block.BlockState;
+import net.minecraft.fluid.FlowableFluid;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.ChunkRegion;
+import net.minecraft.world.chunk.ChunkGenerationSteps;
 import net.minecraft.world.chunk.ChunkStatus;
+import net.minecraft.world.chunk.ProtoChunk;
 import net.minecraft.world.chunk.WorldChunk;
 
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
 
 public class ServerBlockTicking extends NewChunkStatus {
 
@@ -39,6 +52,13 @@ public class ServerBlockTicking extends NewChunkStatus {
 
     @Override
     public CompletionStage<Void> upgradeToThis(ChunkLoadingContext context, Cancellable cancellable) {
+        if (Config.filterFluidPostProcessing) {
+            try {
+                filterFluidTicks(context);
+            } catch (Throwable t) {
+                t.printStackTrace();
+            }
+        }
         return CompletableFuture.runAsync(() -> {
             try (var ignored = ThreadInstrumentation.getCurrent().begin(new ChunkTaskWork(context, this, true))) {
                 final WorldChunk chunk = (WorldChunk) context.holder().getItem().get().chunk();
@@ -50,6 +70,38 @@ public class ServerBlockTicking extends NewChunkStatus {
         }, ((IThreadedAnvilChunkStorage) context.tacs()).getMainThreadExecutor());
     }
 
+    private static void filterFluidTicks(ChunkLoadingContext context) {
+        final WorldChunk chunk = (WorldChunk) context.holder().getItem().get().chunk();
+
+        Supplier<ChunkRegion> chunkRegionSupplier = Suppliers.memoize(() -> new ChunkRegion(((IThreadedAnvilChunkStorage) context.tacs()).getWorld(), context.chunks(), ChunkGenerationSteps.GENERATION.get(ChunkStatus.LIGHT), chunk));
+
+        int total = 0;
+        int eliminated = 0;
+        ShortList[] postProcessingLists = chunk.getPostProcessingLists();
+        for (int i = 0; i < postProcessingLists.length; i++) {
+            if (postProcessingLists[i] != null) {
+                for (ShortListIterator iterator = postProcessingLists[i].iterator(); iterator.hasNext(); ) {
+                    Short short_ = iterator.next();
+                    BlockPos blockPos = ProtoChunk.joinBlockPos(short_, chunk.sectionIndexToCoord(i), chunk.getPos());
+                    BlockState blockState = chunk.getBlockState(blockPos);
+                    FluidState fluidState = blockState.getFluidState();
+                    if (!fluidState.isEmpty() && fluidState.getFluid() instanceof FlowableFluid) {
+                        total ++;
+                        if (!FlowableFluidUtils.needsPostProcessing(chunkRegionSupplier.get(), blockPos, blockState, fluidState)) {
+                            iterator.remove();
+//                            iterator.set((short) (short_ | (0x4000))); // set a flag
+                            eliminated ++;
+                        }
+                    }
+                }
+            }
+        }
+
+//        if (total > 0) {
+//            System.out.println(String.format("Eliminated %d/%d (%.2f%%) post processing fluids in chunk %s", eliminated, total, eliminated / (double) total * 100.0, context.holder().getKey()));
+//        }
+    }
+
     private static void sendChunkToPlayer(ChunkLoadingContext context) {
         final WorldChunk chunk = (WorldChunk) context.holder().getItem().get().chunk();
         CompletableFuture<?> completableFuturexx = context.holder().getUserData().get().getPostProcessingFuture();
