From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Mon, 28 Apr 2025 23:01:41 +0800
Subject: [PATCH] fix: stop ppl hallucination on task thread dump


diff --git a/c2me-base/src/main/java/com/ishland/c2me/base/mixin/report/MixinDedicatedServerWatchdog.java b/c2me-base/src/main/java/com/ishland/c2me/base/mixin/report/MixinDedicatedServerWatchdog.java
index 3afd8cba..c9368791 100644
--- a/c2me-base/src/main/java/com/ishland/c2me/base/mixin/report/MixinDedicatedServerWatchdog.java
+++ b/c2me-base/src/main/java/com/ishland/c2me/base/mixin/report/MixinDedicatedServerWatchdog.java
@@ -37,7 +37,7 @@ public class MixinDedicatedServerWatchdog {
 
     @Inject(method = "run", at = @At(value = "INVOKE", target = "Lnet/minecraft/util/crash/CrashReport;addElement(Ljava/lang/String;)Lnet/minecraft/util/crash/CrashReportSection;", ordinal = 0))
     private void addInstrumentationData(CallbackInfo ci, @Local CrashReport crashReport) {
-        CrashReportSection section = crashReport.addElement("Thread trace dump (obtained on a best-effort basis)");
+        CrashReportSection section = crashReport.addElement("Thread trace dump (obtained on a best-effort basis)", 1);
         try {
             for (Map.Entry<Thread, ThreadState> entry : ThreadInstrumentation.entrySet()) {
                 try {
