From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Mon, 19 May 2025 23:10:50 +0800
Subject: [PATCH] change: cleanup retry stacktraces


diff --git a/c2me-base/src/main/java/com/ishland/c2me/base/common/util/RxJavaUtils.java b/c2me-base/src/main/java/com/ishland/c2me/base/common/util/RxJavaUtils.java
index 69503ec1..de69503d 100644
--- a/c2me-base/src/main/java/com/ishland/c2me/base/common/util/RxJavaUtils.java
+++ b/c2me-base/src/main/java/com/ishland/c2me/base/common/util/RxJavaUtils.java
@@ -16,14 +16,20 @@ public class RxJavaUtils {
         List<Throwable> throwableList = Collections.synchronizedList(new ReferenceArrayList<>());
         throwableList.addAll(List.of(existingErrors));
         return errors -> errors.zipWith(Flowable.range(1, maxRetries + 1), (error, retryCount) -> {
+            throwableList.add(error);
             if (retryCount > maxRetries) {
-                final RuntimeException exception = new RuntimeException("Max retries reached", error);
+                final RuntimeException exception = new StacklessRuntimeException("Max retries reached");
                 throwableList.forEach(exception::addSuppressed);
                 throw exception;
             }
-            throwableList.add(error);
             return retryCount;
         }).flatMap(retryCount -> Flowable.timer((long) Math.pow(2, retryCount - 1) * initialDelayMillis, TimeUnit.MILLISECONDS));
     }
 
+    public static class StacklessRuntimeException extends RuntimeException {
+        public StacklessRuntimeException(String message) {
+            super(message, null, true, false);
+        }
+    }
+
 }
