From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Mon, 26 May 2025 11:01:40 +0800
Subject: [PATCH] new: add elapsed time to task dump


diff --git a/c2me-base/src/main/java/com/ishland/c2me/base/common/threadstate/SyncLoadWork.java b/c2me-base/src/main/java/com/ishland/c2me/base/common/threadstate/SyncLoadWork.java
index 96d72c3e..a930f8b9 100644
--- a/c2me-base/src/main/java/com/ishland/c2me/base/common/threadstate/SyncLoadWork.java
+++ b/c2me-base/src/main/java/com/ishland/c2me/base/common/threadstate/SyncLoadWork.java
@@ -1,14 +1,20 @@
 package com.ishland.c2me.base.common.threadstate;
 
+import com.ishland.c2me.base.common.util.TimeUtil;
 import net.minecraft.server.world.ServerWorld;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.world.chunk.ChunkStatus;
 
-public record SyncLoadWork(ServerWorld world, ChunkPos chunkPos, ChunkStatus targetStatus, boolean create) implements RunningWork {
+public record SyncLoadWork(ServerWorld world, ChunkPos chunkPos, ChunkStatus targetStatus, boolean create, long startTime) implements RunningWork {
+
+    public SyncLoadWork(ServerWorld world, ChunkPos chunkPos, ChunkStatus targetStatus, boolean create) {
+        this(world, chunkPos, targetStatus, create, System.nanoTime());
+    }
 
     @Override
     public String toString() {
-        return String.format("Sync load chunk %s to status %s in world %s (create=%s)", chunkPos, targetStatus, world.getRegistryKey().getValue(), create);
+        return String.format("Sync load chunk %s to status %s in world %s (create=%s) (%s elapsed)",
+                chunkPos, targetStatus, world.getRegistryKey().getValue(), create, TimeUtil.formatElapsedTime(System.nanoTime() - startTime));
     }
 
 }
diff --git a/c2me-base/src/main/java/com/ishland/c2me/base/common/util/TimeUtil.java b/c2me-base/src/main/java/com/ishland/c2me/base/common/util/TimeUtil.java
new file mode 100644
index 00000000..c0be974e
--- /dev/null
+++ b/c2me-base/src/main/java/com/ishland/c2me/base/common/util/TimeUtil.java
@@ -0,0 +1,61 @@
+package com.ishland.c2me.base.common.util;
+
+import java.util.Locale;
+import java.util.concurrent.TimeUnit;
+
+public class TimeUtil {
+
+    public static String formatElapsedTime(long nanos) {
+        if (nanos < 0) {
+            return nanos + " ns";
+        }
+        if (nanos == 0) {
+            return "0 ns";
+        }
+
+        TimeUnit unit = chooseUnit(nanos);
+        double value = (double) nanos / TimeUnit.NANOSECONDS.convert(1, unit);
+
+        return formatCompact4Digits(value) + " " + abbreviate(unit);
+    }
+
+    static String formatCompact4Digits(double value) {
+        return String.format(Locale.ROOT, "%.4g", value);
+    }
+
+    private static TimeUnit chooseUnit(long nanos) {
+        if (TimeUnit.DAYS.convert(nanos, TimeUnit.NANOSECONDS) > 0) {
+            return TimeUnit.DAYS;
+        }
+        if (TimeUnit.HOURS.convert(nanos, TimeUnit.NANOSECONDS) > 0) {
+            return TimeUnit.HOURS;
+        }
+        if (TimeUnit.MINUTES.convert(nanos, TimeUnit.NANOSECONDS) > 0) {
+            return TimeUnit.MINUTES;
+        }
+        if (TimeUnit.SECONDS.convert(nanos, TimeUnit.NANOSECONDS) > 0) {
+            return TimeUnit.SECONDS;
+        }
+        if (TimeUnit.MILLISECONDS.convert(nanos, TimeUnit.NANOSECONDS) > 0) {
+            return TimeUnit.MILLISECONDS;
+        }
+        if (TimeUnit.MICROSECONDS.convert(nanos, TimeUnit.NANOSECONDS) > 0) {
+            return TimeUnit.MICROSECONDS;
+        }
+        return TimeUnit.NANOSECONDS;
+    }
+
+    private static String abbreviate(TimeUnit unit) {
+        return switch (unit) {
+            case NANOSECONDS -> "ns";
+            case MICROSECONDS -> "\u03bcs"; // μs
+            case MILLISECONDS -> "ms";
+            case SECONDS -> "s";
+            case MINUTES -> "min";
+            case HOURS -> "h";
+            case DAYS -> "d";
+            default -> throw new AssertionError();
+        };
+    }
+
+}
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/threadstate/ChunkTaskWork.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/threadstate/ChunkTaskWork.java
index b04b141f..35f0bd55 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/threadstate/ChunkTaskWork.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/threadstate/ChunkTaskWork.java
@@ -1,20 +1,32 @@
 package com.ishland.c2me.rewrites.chunksystem.common.threadstate;
 
 import com.ishland.c2me.base.common.threadstate.RunningWork;
+import com.ishland.c2me.base.common.util.TimeUtil;
 import com.ishland.c2me.base.mixin.access.IThreadedAnvilChunkStorage;
 import com.ishland.c2me.rewrites.chunksystem.common.ChunkLoadingContext;
 import com.ishland.c2me.rewrites.chunksystem.common.NewChunkStatus;
 import net.minecraft.server.world.ServerWorld;
 import net.minecraft.util.math.ChunkPos;
 
-public record ChunkTaskWork(ServerWorld world, ChunkPos chunkPos, NewChunkStatus status, boolean isUpgrade) implements RunningWork {
+public record ChunkTaskWork(ServerWorld world, ChunkPos chunkPos, NewChunkStatus status, boolean isUpgrade, long startTime) implements RunningWork {
 
     public ChunkTaskWork(ChunkLoadingContext context, NewChunkStatus status, boolean isUpgrade) {
         this(
                 ((IThreadedAnvilChunkStorage) context.tacs()).getWorld(),
                 context.holder().getKey(),
                 status,
-                isUpgrade
+                isUpgrade,
+                System.nanoTime()
+        );
+    }
+
+    public ChunkTaskWork(ServerWorld world, ChunkPos chunkPos, NewChunkStatus status, boolean isUpgrade) {
+        this(
+                world,
+                chunkPos,
+                status,
+                isUpgrade,
+                System.nanoTime()
         );
     }
 
@@ -22,17 +34,19 @@ public record ChunkTaskWork(ServerWorld world, ChunkPos chunkPos, NewChunkStatus
     public String toString() {
         if (isUpgrade) {
             return String.format(
-                    "Upgrading chunk %s to %s in world %s",
+                    "Upgrading chunk %s to %s in world %s (%s elapsed)",
                     chunkPos,
                     status,
-                    world.getRegistryKey().getValue()
+                    world.getRegistryKey().getValue(),
+                    TimeUtil.formatElapsedTime(System.nanoTime() - startTime)
             );
         } else {
             return String.format(
-                    "Downgrading chunk %s from %s in world %s",
+                    "Downgrading chunk %s from %s in world %s (%s elapsed)",
                     chunkPos,
                     status,
-                    world.getRegistryKey().getValue()
+                    world.getRegistryKey().getValue(),
+                    TimeUtil.formatElapsedTime(System.nanoTime() - startTime)
             );
         }
     }
