From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Fri, 30 May 2025 17:24:57 +0800
Subject: [PATCH] fix: clear "broken" flag when unloaded but not removed


diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
index 5025e5bd..f474d2ad 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
@@ -164,7 +164,17 @@ public class ReadFromDisk extends NewChunkStatus {
 
                 context.holder().getItem().set(new ChunkState(null, null, null));
             }
-        }, ((IThreadedAnvilChunkStorage) context.tacs()).getMainThreadExecutor());
+        }, ((IThreadedAnvilChunkStorage) context.tacs()).getMainThreadExecutor())
+        .whenComplete((unused, throwable) -> {
+            try {
+                if ((context.holder().getFlags() & ItemHolder.FLAG_BROKEN) != 0) {
+                    LOGGER.warn("Broken chunk {} was unloaded", context.holder().getKey());
+                    context.holder().clearFlag(ItemHolder.FLAG_BROKEN);
+                }
+            } catch (Throwable t) {
+                t.printStackTrace();
+            }
+        });
     }
 
     protected CompletionStage<Void> syncWithLightEngine(ChunkLoadingContext context) {
