From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Wed, 11 Jun 2025 22:33:45 +0800
Subject: [PATCH] new: lowMemoryMode again


diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
index 119ca607..eae73915 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
@@ -91,6 +91,16 @@ public class Config {
                     """)
             .getBoolean(false, false);
 
+    public static final boolean lowMemoryMode = new ConfigSystem.ConfigAccessor()
+            .key("chunkSystem.lowMemoryMode")
+            .comment("""
+                    Whether to enable low memory mode
+                    
+                    This option will attempt to aggressively unload unused chunks.
+                    Only applies when useLegacyScheduling is disabled.
+                    """)
+            .getBoolean(false, false);
+
     public static void init() {
         // intentionally empty
     }
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/NewChunkHolderVanillaInterface.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/NewChunkHolderVanillaInterface.java
index 546d19aa..441e73a1 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/NewChunkHolderVanillaInterface.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/NewChunkHolderVanillaInterface.java
@@ -186,7 +186,7 @@ public class NewChunkHolderVanillaInterface extends ChunkHolder implements IFast
         }
     }
 
-    private void triggerDeferredLoad(NewChunkStatus requestedStatus) {
+    public void triggerDeferredLoad(NewChunkStatus requestedStatus) {
         if (Config.useLegacyScheduling) return;
         synchronized (this) {
             if (this.loadedDeferredStatus != null && this.loadedDeferredStatus.ordinal() >= requestedStatus.ordinal()) {
diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
index f474d2ad..9011f713 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/statuses/ReadFromDisk.java
@@ -71,7 +71,12 @@ public class ReadFromDisk extends NewChunkStatus {
                         }
                     }
                 })
-                .doOnSuccess(chunk -> context.holder().getItem().set(new ChunkState(chunk, chunk, ChunkStatus.EMPTY)))
+                .doOnSuccess(chunk -> {
+                    context.holder().getItem().set(new ChunkState(chunk, chunk, ChunkStatus.EMPTY));
+                    if (!Config.lowMemoryMode) {
+                        context.holder().getUserData().get().triggerDeferredLoad(NewChunkStatus.DISK);
+                    }
+                })
                 .ignoreElement()
                 .cache()
                 .toCompletionStage(null);
