From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Sat, 21 Jun 2025 21:44:34 +0800
Subject: [PATCH] fix: enable legacy scheduling for now


diff --git a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
index eae73915..f0e4a197 100644
--- a/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
+++ b/c2me-rewrites-chunk-system/src/main/java/com/ishland/c2me/rewrites/chunksystem/common/Config.java
@@ -86,10 +86,8 @@ public class Config {
                     Whether to use legacy scheduling for neighbor chunks
                     
                     Enabling this restores the behavior of always loading in neighbor chunks when a chunk is loaded.
-                    
-                    This is currently deprecated and will be removed in the future.
                     """)
-            .getBoolean(false, false);
+            .getBoolean(true, false);
 
     public static final boolean lowMemoryMode = new ConfigSystem.ConfigAccessor()
             .key("chunkSystem.lowMemoryMode")
